lib/teleterm/services/connectmycomputer/connectmycomputer.go

```diff
@@ -0,0 +1,255 @@
+	}
+
+	return &RoleSetup{cfg: cfg}, nil
+}
+
+type RoleSetupResult struct {
+	CertsReloaded bool
+}
+
+// Run ensures that at the end of its execution the user has their own individual Connect My
+// Computer role and that the role includes the current system username in allowed logins.
+//
+// If the role list of the user got updated, the return value has CertsReloaded set to true.
+func (s *RoleSetup) Run(ctx context.Context, accessAndIdentity AccessAndIdentity, certManager CertManager, cluster *clusters.Cluster) (RoleSetupResult, error) {
+	noCertsReloaded := RoleSetupResult{}
+	if !cluster.URI.IsRoot() {
+		return noCertsReloaded, trace.BadParameter("Connect My Computer works only with root clusters")
+	}
+
+	// Do not use GetCurrentUser – it returns the current view of the user given the certs, not merely
+	// the resource from the backend. This means that GetCurrentUser includes any roles granted
+	// through access requests. We don't want that since we're later going to update the user.
+	clusterUser, err := accessAndIdentity.GetUser(cluster.GetLoggedInUser().Name, false /* withSecrets */)
+	if err != nil {
+		return noCertsReloaded, trace.Wrap(err)
+	}
+
+	isLocalUser := clusterUser.GetCreatedBy().Connector == nil
+	if !isLocalUser {
+		return noCertsReloaded,
+			trace.BadParameter("Connect My Computer works only with local users, the user %v was created by %v connector",
+				clusterUser.GetName(), clusterUser.GetCreatedBy().Connector.Type)
+	}
+
+	roleName := fmt.Sprintf("%v%v", teleport.ConnectMyComputerRoleNamePrefix, clusterUser.GetName())
+
+	doesRoleExist := true
+	existingRole, err := accessAndIdentity.GetRole(ctx, roleName)
+	if err != nil {
+		if trace.IsNotFound(err) {
+			doesRoleExist = false
+		} else {
+			return noCertsReloaded, trace.Wrap(err)
+		}
+	}
+
+	systemUser, err := user.Current()
+	if err != nil {
+		return noCertsReloaded, trace.Wrap(err)
+	}
+
+	if !doesRoleExist {
+		s.cfg.Log.Infof("Creating the role %v.", roleName)
+
+		role, err := types.NewRole(roleName, types.RoleSpecV6{
+			Allow: types.RoleConditions{
+				NodeLabels: types.Labels{
+					types.ConnectMyComputerNodeOwnerLabel: []string{clusterUser.GetName()},
+				},
+				Logins: []string{systemUser.Username},
+			},
+		})
+		if err != nil {
+			return noCertsReloaded, trace.Wrap(err)
+		}
+		if err = accessAndIdentity.UpsertRole(ctx, role); err != nil {
+			return noCertsReloaded, trace.Wrap(err, "creating role %v", role.GetName())
+		}
+	} else {
+		if slices.Contains(clusterUser.GetLogins(), systemUser.Username) {
+			s.cfg.Log.Infof("The role %v already exists and includes current system username.", roleName)
+		} else {
+			s.cfg.Log.Infof("Adding %v to the logins of the role %v.", systemUser.Username, roleName)
+
+			allowedLogins := existingRole.GetLogins(types.Allow)
+			existingRole.SetLogins(types.Allow, append(allowedLogins, systemUser.Username))
+
+			timeoutCtx, cancel := context.WithTimeout(ctx, 15*time.Second)
+			defer cancel()
```

Comment:
It's also enforced by our linter. If I don't call cancel I get this:

> lostcancel: the cancel function returned by context.WithTimeout should be called, not discarded, to avoid a context leak (govet)